### PR TITLE
bluetooth scan interface added

### LINKED ISSUE #1845: Add interface to scan bluetooth devices for feature of bluetooth communication with PSLab
**Actual Behaviour**
Currently PSLab can communicate only through wired connection. New feature of communication over bluetooth needs to be added

**Expected Behaviour**
Add interface so user can scan available bluetooth devices.

**Would you like to work on the issue?**
Yes

### PR DESCRIPTION
**Pull request by neel1998**
Fixes #1845 

**Changes**: user can scan available bluetooth devices

**Screenshot/s for the changes**: 

**Checklist**: [Please tick following check boxes with `[x]` if the respective task is completed]
- [x] I have used resources from `strings.xml`, `dimens.xml` and `colors.xml` without hard-coding them
- [x] No modifications done at the end of resource files `strings.xml`, `dimens.xml` or `colors.xml`
- [x] I have reformatted code in every file included in this PR [<kbd>CTRL</kbd>+<kbd>ALT</kbd>+<kbd>L</kbd>]
- [x] My code does not contain any extra lines or extra spaces
- [x] I have requested reviews from other members

**APK for testing**: 

[bluetooth_scan.zip](https://github.com/fossasia/pslab-android/files/3367188/bluetooth_scan.zip)

@mariobehling @CloudyPadmal @cweitat this is the first step to add bluetooth communication to PSLab. Please guide me for further steps. I am willing to work on firmware code as well to add bluetooth communication

### CODE REVIEW

File: app/build.gradle [hunk 1]
```diff
@@ -70,6 +70,7 @@ dependencies {
     implementation 'com.github.Triggertrap:SeekArc:v1.1'
 
     implementation "com.jakewharton:butterknife:$rootProject.butterKnifeVersion"
+    implementation 'com.android.support.constraint:constraint-layout:1.1.3'
```
Comment (CloudyPadmal) [inline app/build.gradle]: Duplicate. Check line 48

**Comment by neel1998** [comment]
> User cannot connect with both USB and bluetooth at the same time. Use the "USB" icon in menubar to connect device with.
> 
> 1. If USB connection is present, connect with USB
> 2. If USB connection is not present, ask user permission using a dialog box to scan for bluetooth connections and then do the scan and connect. Don't direct user to a different activity. Show the device scan result in a list view dialog box. We can later expand this option to WiFi as well.

Great suggestion. Will surely make the change

**Comment by neel1998** [comment]
@CloudyPadmal I have made the changes you suggested. Please have a look. Thanks.

File: app/build.gradle (hunk 1 above)
Comment (CloudyPadmal) [reply app/build.gradle]: This is still here

File: app/src/main/res/layout/fragment_bluetooth_scan.xml [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+    <Button
+        android:id="@+id/bluetooth_scan_stop_button"
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:layout_centerHorizontal="true"
+        android:background="@color/colorPrimary"
+        android:text="@string/bluetooth_stop_text"
+        android:textColor="@color/white"
+        android:layout_margin="10dp"
```
Comment (CloudyPadmal) [inline app/src/main/res/layout/fragment_bluetooth_scan.xml]: Hardcoded dimensions

**Comment by neel1998** [comment]
@CloudyPadmal I will make the changes. Can you please guide me for the next steps. Thanks

**Comment by neel1998** [comment]
@CloudyPadmal  check the updated apk and screenshot. Thanks

File: app/src/main/java/io/pslab/fragment/BluetoothScanFragment.java [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+    public void onResume() {
+        super.onResume();
+        ViewGroup.LayoutParams params = getDialog().getWindow().getAttributes();
+        params.height = ViewGroup.LayoutParams.MATCH_PARENT;
+        params.width = ViewGroup.LayoutParams.MATCH_PARENT;
+        getDialog().getWindow().setAttributes((android.view.WindowManager.LayoutParams) params);
+    }
+
+    private void connectBluetooth() {
+        Log.d(BLT_TAG, "connect called");
```
Comment (CloudyPadmal) [inline app/src/main/java/io/pslab/fragment/BluetoothScanFragment.java]: Remove this

File: app/src/main/java/io/pslab/fragment/BluetoothScanFragment.java [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+                }
+            },0, 2000);
+            readData();
+        } catch (IOException e) {
+            Log.d(BLT_TAG, "connection failed");
+            e.printStackTrace();
+        }
+    }
+    private void sendData(byte data) {
+        Log.d(BLT_TAG, "send called");
```
Comment (CloudyPadmal) [inline app/src/main/java/io/pslab/fragment/BluetoothScanFragment.java]: Remove debugging logs

File: app/src/main/java/io/pslab/fragment/BluetoothScanFragment.java [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+        try {
+            mOutputStream.write(data);
+            Log.d(BLT_TAG, "data sent");
+        } catch (IOException e) {
+            Log.d(BLT_TAG, "send failed");
```
Comment (CloudyPadmal) [inline app/src/main/java/io/pslab/fragment/BluetoothScanFragment.java]: Can put error stacktrace to log instead

File: app/src/main/res/layout/fragment_bluetooth_scan.xml [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+        android:id="@+id/bluetooth_scanned_devices_list"
+        android:layout_width="match_parent"
+        android:layout_height="wrap_content"
+        android:layout_below="@+id/scanned_devices_list_title"
+        android:background="@drawable/rectangle_border_black"
+        android:layout_marginTop="@dimen/bluetooth_dialog_margin"
+        android:divider="@color/dark_grey"
+        android:dividerHeight="@dimen/bluetooth_list_divider_height"
+        android:padding="@dimen/bluetooth_dialog_margin"
+        />
```
Comment (CloudyPadmal) [inline app/src/main/res/layout/fragment_bluetooth_scan.xml]: Format these. Let's submit a quality code for review

**Comment by neel1998** [comment]
@CloudyPadmal this is a WIP . That's why I kept the log.

**Comment by neel1998** [comment]
@CloudyPadmal should i remove the changes from last commit ad make seperate PR for conection and reading/writing so that, the scan interface PR can get merged?

**Comment by CloudyPadmal** [comment]
Yes please do so

**Comment by neel1998** [comment]
@CloudyPadmal  i have removed the connection/read/write code, please review and merge. Thanks

File: app/src/main/res/layout/fragment_bluetooth_scan.xml (hunk 6 above)
Comment (CloudyPadmal) [reply app/src/main/res/layout/fragment_bluetooth_scan.xml]: These are still here

**Comment by neel1998** [comment]
@CloudyPadmal  made the changes.

**Comment by neel1998** [comment]
@CloudyPadmal  can we please merge this?

**Pull request merged**